### PR TITLE
Base Type Refactors

### PR DESCRIPTION
**Pull request by mcook42**
## Summary
This Pull Request refactors multiple classes in the codebase to improve modularity and ensure better adherence to ASN.1 specifications.

### Main Changes:
- Added default interface method `hasExtensionMarker()` to Asn1Type to reduce boilerplate code and increase usage of composition over inheritance. Generated classes with an extensibility marker include an override method to indicate they are extensible.
- Refactored Asn1Sequence, Asn1SequenceOf, and related types in the runtime module for better documentation and functionality.
- Enhanced Asn1SequenceOf to extend compatibility with Java's List interface, enforcing constraints like size validation.
- Added a comprehensive set of unit tests for Asn1SequenceOf to validate functionality, size constraints, and behavior under edge cases.

## Motivation

Since these models will underpin much of the jpo-ode, it's important we define the interfaces appropriately before we implement all of the models and add them into the core jpo-ode platform. Composition over inheritance helps us adhere more to the Open–closed principle.

### CODE REVIEW

File: jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1Sequence.java [hunk 1]
```diff
@@ -1,36 +1,24 @@
...
+ *   <li>Elements may be of different types</li>
+ *   <li>The structure is statically defined (except for optional elements)</li>
+ * </ul>
+ * <p>
+ *
+ * @see Asn1Type The base interface for all ASN.1 types
+ * @since 1.0
  */
-@Slf4j
-@ToString
```
Comment (mcook42) [inline jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1Sequence.java:10]: Note: an abstract base class without any properties doesn't need a logger or a toString method since it cannot be instantiated directly

File: jpo-asn-runtime/src/test/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOfTest.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        // Third item should throw exception
+        IllegalStateException exception = assertThrows(IllegalStateException.class, () -> {
+            sequence.add(item3);
+        });
+
+        assertTrue(exception.getMessage().contains("exceed maximum allowed size"));
+    }
+
+    @Test
+    void testAddAsn1TypeExceedingUpperBound() {
```
Comment (payneBrandon) [inline jpo-asn-runtime/src/test/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOfTest.java]: question: is this a duplicate of `testAddExceedingUpperBound`?

File: jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOf.java [hunk 3]
```diff
@@ -1,43 +1,306 @@
...
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+import java.util.function.UnaryOperator;
 
 /**
  * Base class for ASN.1 SEQUENCE-OF types.
- * @param <T> The type of items in the SEQUENCE-OF
+ * <p>
```
Comment (payneBrandon) [inline jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOf.java:12]: praise: great description here!

File: jpo-asn-runtime/src/test/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOfTest.java (hunk 2 above)
Comment (mcook42) [reply jpo-asn-runtime/src/test/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOfTest.java]: it is. good catch. I've removed it https://github.com/CDOT-CV/jpo-asn-pojos/pull/14/commits/d450d0b4daa81800feafc3b8b69ed7eb4b753e9f

**Review by payneBrandon — APPROVED** [review]
lgtm!

File: jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1Type.java [hunk 4]
```diff
@@ -1,11 +1,32 @@
...
+     *      *    }
+     *      * </pre>
+     * *
+     *
+     * @return true if the ASN.1 type has an extension marker, false otherwise
+     */
+    @JsonIgnore
+    default boolean hasExtensionMarker() {
+        return false;
+    }
```
Comment (iyourshaw) [inline jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1Type.java:31]: I like this.  Moving the extension marker to the root interface makes a lot of sense, since all the types are potentially extensible.

File: jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOf.java [hunk 5]
```diff
@@ -1,43 +1,306 @@
...
+ *
+ * @param <T> The type of items in the SEQUENCE OF, must extend {@link Asn1Type}
+ *
+ * @see Asn1Type The base interface for all ASN.1 types
+ * @see Asn1Sequence For representing ASN.1 SEQUENCE types with named fields
  */
-public abstract class Asn1SequenceOf<T extends Asn1Type>
-    extends ArrayList<T>
-    implements Asn1Type {
+public abstract class Asn1SequenceOf<T extends Asn1Type> implements Asn1Type, List<T> {
```
Comment (iyourshaw) [inline jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOf.java]: I get that composition and "programming to the interface" is generally preferable and has the advantage that in theory it would be easier to change the underlying list implementation, so this will prevent clients from depending on the underlying implementation.   But there is quite a lot of code here that only wraps the `ArrayList` methods.  Please consider instead extending [AbstractList](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/AbstractList.html) which is intended for exactly this sort of thing, as an alternative to extending ArrayList. Abstract list would take care of most of the boilerplate, and also is the bare minimum interface that is needed.  The items in the collection must be ordered and randomly accessible, so it seems very unlikely we would ever want to switch out the implementation to something that implements the List interface, but doesn't also inherit from AbstractList.

**Review by iyourshaw — CHANGES_REQUESTED** [review]
I like the ideas here.  One recommendation on simplifying the SequenceOf implementation.  I will also try regenerating the classes with these runtime changes.  There are a number of special cases related to SEQUENCE-OF various types, which don't have any examples in the PR, so will need to verify that all the tests in the larger project still work.

File: jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOf.java (hunk 5 above)
Comment (mcook42) [reply jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1SequenceOf.java]: Great idea. I've updated the code to use the AbstractList https://github.com/CDOT-CV/jpo-asn-pojos/pull/14/commits/e013f45aa43d950121c77c8baf49bac7d09372e0

**Review by iyourshaw — APPROVED** [review]
looks good, excellent documentation!
